### PR TITLE
[DEV-69] Basic RAFT NFT

### PR DESCRIPTION
**Pull request by grp06**

### CODE REVIEW

File: src/Raft.sol [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    Counters.Counter private _tokenIds;
+
+    constructor(address owner, string memory name, string memory symbol) ERC721(name, symbol) {
+      transferOwnership(owner);
+      _pause();
+    }
+
+    function mint(address recipient, string memory tokenURI) external returns (uint) { 
+        require(msg.sender == owner() || !paused(), "mint: unauthorized to mint");
+        _tokenIds.increment();
```
[the remark this answers is not in the document]
Comment (grp06) [reply src/Raft.sol:22]: No, basically what I'm saying is "Owner can mint whenever he wants" ... "non-owners can mint as long as it's not paused"

File: src/Raft.sol [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity ^0.8.3;
+
+import "@openzeppelin/contracts/utils/Counters.sol";
+import "@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/security/Pausable.sol";
+
+contract Raft is ERC721URIStorage, Ownable, Pausable {
```
Comment (rahulrumalla) [inline src/Raft.sol:11]: I think it's time we start documenting full-fledged comments using natspec 
https://docs.soliditylang.org/en/v0.8.15/natspec-format.html

Comment (rahulrumalla) [reply src/Raft.sol:11]: Let's add inline comments too to document why we're transferring ownership in ctor or setting "pause" for example

File: test/raft.test.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        return { Raft, raftContract, owner, addr1 }
+    }
+
+    it('should deploy the contract with the right params', async function () {
+        const { raftContract } = await loadFixture(deployContractFixture)
+        const deployedContractName = await raftContract.name()
+        const deployedSymbolName = await raftContract.symbol()
+        expect(deployedContractName).to.equal(name)
+        expect(deployedSymbolName).to.equal(symbol)
+
```
Comment (rahulrumalla) [inline test/raft.test.ts:29]: Here you should also check the owner of the contract and the paused state as these are being set in the constructor

File: test/raft.test.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import { loadFixture } from '@nomicfoundation/hardhat-network-helpers'
+
+describe('Raft', async function () {
+    let raftContract: Raft
+    const name = 'Raft'
+    const symbol = 'RAFT'
+    const chainId = 31337
+    const tokenURI = 'blah'
+
+    async function deployContractFixture() {
```
Comment (rahulrumalla) [inline test/raft.test.ts:15]: why not leverage `before()`?

File: test/raft.test.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        expect(deployedSymbolName).to.equal(symbol)
+
+        const provider = waffle.provider
+        const network = await provider.getNetwork()
+        const deployedToChainId = network.chainId
+        expect(deployedToChainId).to.equal(chainId)
+    })
+
+    it('should allow owner to mint when minting is paused', async function () {
+        const { raftContract, addr1 } = await loadFixture(deployContractFixture)
```
Comment (rahulrumalla) [inline test/raft.test.ts:37]: How about we add explicit assertion that the assumed state of the contract is as-needed in order for the test to pass?

for ex,
```typescript
expect(isPaused).to.equal(false)
```

**Review by rahulrumalla — CHANGES_REQUESTED** [review]
Left some comments. Good stuff @grp06 :D

File: test/raft.test.ts (hunk 4 above)
Comment (grp06) [reply test/raft.test.ts:15]: I found this method in the Hardhat testing docs here: https://hardhat.org/tutorial/testing-contracts#reusing-common-test-setups-with-fixtures

File: src/Raft.sol (hunk 2 above)
Comment (grp06) [reply src/Raft.sol:15]: done

File: test/raft.test.ts (hunk 3 above)
Comment (grp06) [reply test/raft.test.ts:30]: done

**Review by rahulrumalla — APPROVED** [review]

**Pull request merged**